syntax = "proto3";

package ai.verta.common;
option go_package = "github.com/VertaAI/modeldb/protos/gen/go/protos/public/common";

option java_multiple_files = true;
import "google/protobuf/struct.proto";

/*
 * All proto common entity
 */

message TernaryEnum {
    enum Ternary {
        UNKNOWN = 0;
        TRUE = 1;
        FALSE = 2;
    }
}

/*
 * All proto common entity
 */

message KeyValue {
    string key = 1;
    google.protobuf.Value value = 2;

    /*
    *   As per documentation of proto buffer 3.
    *   For enums, the default value is the first defined enum value, which must be 0.
    */
    ValueTypeEnum.ValueType value_type = 3;
}

message ValueTypeEnum {
    enum ValueType {
        STRING = 0;
        NUMBER = 1;
        LIST = 2;
        BLOB = 3;
    }
}

message CollaboratorTypeEnum {
    enum CollaboratorType {
        READ_ONLY = 0;
        READ_WRITE = 1;
    }
}

message CollaboratorTypeEnumWithDefaultEnum {
    enum CollaboratorTypeEnumWithDefault {
        UNKNOWN = 0;
        READ_WRITE = 1;
        READ_ONLY = 2;
    }
}

message EntitiesEnum {
    enum EntitiesTypes {
        UNKNOWN = 0; // Resource 0 should always cause an error
        ORGANIZATION = 1;
        TEAM = 2;
        USER = 3;
    }
};

message ModelDBResourceEnum {
    enum ModelDBServiceResourceTypes {
        UNKNOWN = 0; // Resource 0 should always cause an error
        ALL = 1;
        PROJECT = 2;
        EXPERIMENT = 3;
        EXPERIMENT_RUN = 4;
        DATASET = 5;
        DATASET_VERSION = 6;
        DASHBOARD = 7;
        REPOSITORY = 8;
        REGISTERED_MODEL = 9;
        REGISTERED_MODEL_VERSION = 10;
        MONITORED_ENTITY = 11;
        NOTIFICATION_CHANNEL = 12;
        ENDPOINT = 13;
        BUILD = 14;
        ORGANIZATION = 15;
        TEAM = 16;
        WORKSPACE = 17;
        USER = 18;
        SERVICE_ACCOUNT = 19;
        CONTAINER_REGISTRY_CONFIGURATION = 20;
        STAGE = 21;
        PROFILER = 22;
    }
};

// For pagination
message Pagination {
    int32 page_number = 2;
    int32 page_limit = 3;
}

message WorkspaceTypeEnum {
    enum WorkspaceType {
        UNKNOWN = 0;
        ORGANIZATION = 1;
        USER = 2;
    }
}
message ArtifactTypeEnum{
    enum ArtifactType {
        IMAGE = 0;
        MODEL = 1;
        TENSORBOARD = 2;
        DATA = 3;
        BLOB = 4;
        STRING = 5;
        CODE = 6;
        CONTAINER = 7;
    }
}

message Artifact {
    string key = 1;
    string path = 2;
    bool path_only = 3;
    ArtifactTypeEnum.ArtifactType artifact_type = 4;
    string linked_artifact_id = 5;
    string filename_extension = 6;
    string serialization = 7;
    string artifact_subtype = 8;
    bool upload_completed = 9;
}

message KeyValueQuery {
    string key = 1; //entity field key
    google.protobuf.Value value = 2; //entity field value
    common.ValueTypeEnum.ValueType value_type = 3; // Default = STRING
    OperatorEnum.Operator operator = 4; // Default = EQ
}

message ArtifactPart {
    uint64 part_number = 1;
    string etag = 2;  // checksum returned by S3
}

message OperatorEnum {
    enum Operator {
        EQ = 0;
        NE = 1;
        GT = 2;
        GTE = 3;
        LT = 4;
        LTE = 5;
        CONTAIN = 6;
        NOT_CONTAIN = 7;
        IN = 8; // This is not fully supported by backend.
        NOT_IN = 9;
    }
}

message VisibilityEnum {
    enum Visibility {
        //Default to private
        PRIVATE = 0;
        //Not supported
        PUBLIC = 1;
        //Public to organization
        ORG_SCOPED_PUBLIC = 2;
        //Default access for organization
        ORG_DEFAULT = 3;
    }
}

// code version
message CodeVersion {
    oneof code {
        GitSnapshot git_snapshot = 1;
        common.Artifact code_archive = 2; // ZIP archive of source code
    }
    uint64 date_logged = 3;
}

message GitSnapshot {
    repeated string filepaths = 1; // paths to relevant source code
    string repo = 2; // URL to remote repository
    string hash = 3; // commit hash
    common.TernaryEnum.Ternary is_dirty = 4;
}

message Empty {

}

message GroupId {
    string org_id = 1;
    string group_id = 2;
}
